### PR TITLE
Code to enable netflow on Openstack setups

### PR DESCRIPTION
**Pull request by tanyatukade**
Added code to enable netflow on openstack setups

### CODE REVIEW

File: aim/db/models.py [hunk 1]
```diff
@@ -32,6 +32,13 @@ class Tenant(model_base.Base, model_base.HasDisplayName,
                       model_base.to_tuple(model_base.Base.__table_args__))
 
 
+class Infra(model_base.Base, model_base.AttributeMixin,
+            model_base.HasAimId, model_base.HasName):
+    __tablename__ = 'aim_infra_netflow'
```
Comment (kentwu1) [inline aim/db/models.py:37]: This should be just aim_infra as its not specific to netflow.

File: aim/db/migration/alembic_migrations/versions/4caed435b0cd_add_netflow.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+        sa.Column('monitored', sa.Boolean, nullable=False, default=False),
+        sa.Column('epoch', sa.BigInteger(), nullable=False,
+                  server_default='0'),
+        sa.PrimaryKeyConstraint('aim_id'),
+        sa.UniqueConstraint('name',
+                            name='uniq_aim_netflow_exporter_pol_identity'),
+        sa.Index('idx_aim_netflow_exporter_pol_identity', 'name'))
+
+    op.create_table(
+        'aim_infra_netflow',
```
Comment (kentwu1) [inline aim/db/migration/alembic_migrations/versions/4caed435b0cd_add_netflow.py:53]: This should be just aim_infra as its not specific to netflow.

File: aim/db/migration/alembic_migrations/versions/4caed435b0cd_add_netflow.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+import sqlalchemy as sa
+
+
+def upgrade():
+    op.create_table(
+        'aim_netflow_exporter_pol',
+        sa.Column('aim_id', sa.String(64), nullable=False),
+        sa.Column('name', sa.String(64), nullable=False),
+        sa.Column('display_name', sa.String(256), nullable=False, default=''),
+        sa.Column('dst_addr', sa.String(64)),
```
Comment (kentwu1) [inline aim/db/migration/alembic_migrations/versions/4caed435b0cd_add_netflow.py:40]: In the model, all these 4 fields only require 16 bytes however here it uses 64.....We should make it consistent then.

File: aim/tests/unit/agent/aid_universes/test_converter.py [hunk 4]
```diff
@@ -213,6 +213,27 @@ class TestAciToAimConverterBD(TestAciToAimConverterBase, base.TestAimDBBase):
         l2_unknown_unicast_mode='flood')
 
 
+class TestAciToAimConverterNetflow(TestAciToAimConverterBase,
+                                   base.TestAimDBBase):
+    resource_type = resource.NetflowVMMExporterPol
+    reverse_map_output = [{'exceptions': {},
+                           'resource': 'netflowVmmExporterPol'}]
+    sample_input = [base.TestAimDBBase._get_example_aci_netflow(),
+                    base.TestAimDBBase._get_example_aci_netflow(
```
Comment (kentwu1) [inline aim/tests/unit/agent/aid_universes/test_converter.py:222]: We should use this one to overwrite those default parameters in the _get_example_aci_netflow().

File: aim/tests/unit/agent/aid_universes/test_converter.py [hunk 5]
```diff
@@ -3923,6 +3944,30 @@ class TestAimToAciConverterVmmInjHost(TestAimToAciConverterBase,
     ]
 
 
+class TestAimToAciConverterNetflow(TestAimToAciConverterBase,
+                                   base.TestAimDBBase):
+    sample_input = [
+        base.TestAimDBBase._get_example_aim_netflow(),
+        base.TestAimDBBase._get_example_aim_netflow(name='netflow')]
```
Comment (kentwu1) [inline aim/tests/unit/agent/aid_universes/test_converter.py:3951]: We should use this one to overwrite those default parameters in the _get_example_aim_netflow().

File: aim/tests/unit/test_aim_manager.py [hunk 6]
```diff
@@ -713,6 +712,42 @@ class TestBridgeDomainMixin(object):
...
+    test_identity_attributes = {'name': 'netflow1'}
+    test_required_attributes = {'name': 'netflow1',
+                                'dst_addr': '172.28.184.76',
+                                'dst_port': '2055',
+                                'src_addr': '1.2.2.2',
+                                'ver': 'v9'}
+    test_search_attributes = {'name': 'netflow1'}
+    test_update_attributes = {'dst_addr': '172.28.184.77',
+                              'dst_port': '2056',
+                              'src_addr': '1.1.2.2',
```
Comment (kentwu1) [inline aim/tests/unit/test_aim_manager.py:739]: we should make src_addr and ver different also here.

File: aim/db/models.py (hunk 1 above)
Comment (tanyatukade) [reply aim/db/models.py]: addressed

File: aim/db/migration/alembic_migrations/versions/4caed435b0cd_add_netflow.py (hunk 2 above)
Comment (tanyatukade) [reply aim/db/migration/alembic_migrations/versions/4caed435b0cd_add_netflow.py]: addressed

File: aim/db/migration/alembic_migrations/versions/4caed435b0cd_add_netflow.py (hunk 3 above)
Comment (tanyatukade) [reply aim/db/migration/alembic_migrations/versions/4caed435b0cd_add_netflow.py]: addressed

File: aim/tests/unit/test_aim_manager.py (hunk 6 above)
Comment (tanyatukade) [reply aim/tests/unit/test_aim_manager.py]: addressed all the comments in the latest commit

File: aim/db/models.py [hunk 7]
```diff
@@ -83,6 +90,20 @@ def to_attr(self, session):
...
 
+class NetflowVMMExporterPol(model_base.Base, model_base.HasDisplayName,
+                            model_base.HasAimId, model_base.AttributeMixin,
+                            model_base.IsMonitored, model_base.HasName):
+    """DB model for Netflow VMM Exporter"""
+    __tablename__ = 'aim_netflow_exporter_pol'
+    __table_args__ = (model_base.uniq_column(__tablename__, 'name') +
+                      model_base.to_tuple(model_base.Base.__table_args__))
+
+    dst_addr = sa.Column(sa.String(16))
```
Comment (kentwu1) [inline aim/db/models.py:101]: is 16 bytes enough for the dest_addr and src_addr? Do we have to consider IPV6 address here?

Comment (tanyatukade) [reply aim/db/models.py]: Its safe to support ipv6 as well. Made changes to support it in the latest commit.

File: aim/db/migration/alembic_migrations/versions/4caed435b0cd_add_netflow.py [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+from alembic import op
+import sqlalchemy as sa
+
+
+def upgrade():
+    op.create_table(
+        'aim_netflow_exporter_pol',
+        sa.Column('aim_id', sa.String(64), nullable=False),
+        sa.Column('name', sa.String(64), nullable=False),
+        sa.Column('display_name', sa.String(256), nullable=False, default=''),
```
Comment (tbachman) [inline aim/db/migration/alembic_migrations/versions/4caed435b0cd_add_netflow.py:39]: We probably can make this just 64 bytes - according to this link, the nameAlias is 0-63 bytes:
https://pubhub.devnetcloud.com/media/apic-mim-ref-501/docs/MO-netflowVmmExporterPol.html

File: aim/db/migration/alembic_migrations/versions/4caed435b0cd_add_netflow.py [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+        sa.Column('dst_addr', sa.String(64)),
+        sa.Column('dst_port', sa.String(16)),
+        sa.Column('src_addr', sa.String(64)),
+        sa.Column('ver', sa.String(16)),
```
Comment (tbachman) [inline aim/db/migration/alembic_migrations/versions/4caed435b0cd_add_netflow.py:43]: ver is just an Enum8 in ACI with 3 values. That said, I can't find any examples in AIM of any String being less than 16 bytes, so this is okay.

File: aim/db/migration/alembic_migrations/versions/4caed435b0cd_add_netflow.py (hunk 8 above)
Comment (tanyatukade) [reply aim/db/migration/alembic_migrations/versions/4caed435b0cd_add_netflow.py]: Addressed in the latest commit.

Comment (tanyatukade) [reply aim/db/migration/alembic_migrations/versions/4caed435b0cd_add_netflow.py:39]: Switched it back to 256 bytes as per docs

**Comment by tbachman** [comment]
Merged - congrats on your first patch! :-)

**Pull request merged**